src/core/libmaven/isotopeDetection.cpp

```diff
@@ -63,207 +63,182 @@ void IsotopeDetection::pullIsotopes(PeakGroup* parentgroup)
-    map<string, PeakGroup> isotopes;
-
-    for (unsigned int s = 0; s < _mavenParameters->samples.size(); s++) {
-        mzSample* sample = _mavenParameters->samples[s];
-        for (unsigned int k = 0; k < masslist.size(); k++) {
-            //			if (stopped())
-            //				break; TODO: stop
-            Isotope& x = masslist[k];
-            string isotopeName = x.name;
-            double isotopeMass = x.mass;
-            double expectedAbundance = x.abundance;
-
-            float mzmin = isotopeMass -_mavenParameters->compoundMassCutoffWindow->massCutoffValue(isotopeMass);
-            float mzmax = isotopeMass +_mavenParameters->compoundMassCutoffWindow->massCutoffValue(isotopeMass);
-
-            float rt = parentgroup->medianRt();
-            float rtmin = parentgroup->minRt;
-            float rtmax = parentgroup->maxRt;
-
-            Peak* parentPeak = parentgroup->getPeak(sample);
-            if (parentPeak) {
-                rt = parentPeak->rt;
-                rtmin = parentPeak->rtmin;
-                rtmax = parentPeak->rtmax;
-            }
-
-            float isotopePeakIntensity = 0;
-            float parentPeakIntensity = 0;
-
-            if (parentPeak) {
-                parentPeakIntensity = parentPeak->peakIntensity;
-                Scan* scan = parentPeak->getScan();
-                std::pair<float, float> isotope = getIntensity(scan, mzmin, mzmax);
-                isotopePeakIntensity = isotope.first;
-                rt = isotope.second;
-            }
+    map<string, PeakGroup> isotopeGroups;
+    for (auto sample : _mavenParameters->samples) {
+        for (Isotope& isotope : masslist) {
+            string isotopeName = isotope.name;
+            float expectedAbundance = static_cast<float>(isotope.abundance);
+
+            float isotopeMass = static_cast<float>(isotope.mass);
+            float mzDelta = _mavenParameters->compoundMassCutoffWindow->massCutoffValue(isotopeMass);
+            float mzmin = isotopeMass - mzDelta;
+            float mzmax = isotopeMass + mzDelta;
+
+            Peak* parentPeak = parentGroup->getPeak(sample);
+            // we should not attempt to find isotopic peaks for samples in
+            // which there is no parent peak
+            if (parentPeak == nullptr)
+                continue;
 
-            if (isotopePeakIntensity == 0 || rt == 0) continue;
+            // even though we will do a peak detection on the isotope's slice
+            // later on, this is a good fast check to see if there is anything
+            // worth searching for; maybe also for adduct & fragment detection?
+            Scan* scan = parentPeak->getScan();
+            std::pair<float, float> topMatch = getIntensity(scan, mzmin, mzmax);
+            float isotopePeakIntensity = topMatch.first;
+            float rt = topMatch.second;
 
-            if (filterIsotope(x, isotopePeakIntensity, parentPeakIntensity, sample, parentgroup))
```

Comment:
Would suggest moving the correlation check to this point? Considering you have the isotope's mzmin and mzmax as well as the parent peak's RT range already. The speed might not take as much of a hit. I think the same correlation calculation is happening multiple times right now.

Reply:
Good point. Will shift.